framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1108,54 +1194,140 @@ int pcm_stop(struct pcm *pcm)
 	if (pcm == NULL) {
 		return -EINVAL;
 	}
+
+	if (!pcm->running) {
+		return oops(pcm, EINVAL, "PCM already in stop state");
+	}
+
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	if (ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
 #else
 	if (ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
 #endif
 		return oops(pcm, errno, "cannot stop channel");
 
+	/* Remove any pending messages from the message queue */
+	struct timespec st_time;
+	struct audio_msg_s msg;
+	unsigned int size;
+	int prio;
+	while (1) {
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+		if (size != sizeof(msg)) {
+			break;
+		}
+	}
+
 	pcm->prepared = 0;
 	pcm->running = 0;
-	pcm->bufPtr = 0;
-	pcm->nextSize = 0;
-	pcm->nextOffset = 0;
-	pcm->nextBuf = NULL;
+	pcm->draining = 0;
+	pcm->buf_ptr = 0;
+	pcm->next_size = 0;
+	pcm->next_offset = 0;
+	pcm->next_buf = NULL;
 
 	return 0;
 }
 
+/** Stops a PCM. Any data present in the buffers will be dropped.
+ * @param pcm A PCM handle.
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drop(struct pcm *pcm)
+{
+	return pcm_stop(pcm);
+}
+
+/** Stop a PCM preserving pending frames.
+ * @param pcm A PCM handle.
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drain(struct pcm *pcm)
+{
+	if (pcm == NULL) {
+		return -EINVAL;
+	}
+
+	if (!pcm->running) {
+		return oops(pcm, EINVAL, "PCM is already stopped.");
+	}
+
+	if (pcm->flags & PCM_OUT) {
```

Comment:
I think we should change pcm->flags & PCM_OUT to !(pcm->flags & PCM_IN) because PCM_OUT is defined as 0 and bitwise & is always zero. So the if condition is not true when pcm is opened in PCM_OUT mode.

Reply:
I modified the flags as below so that such problem does not happen when the flags are used in future:

#define PCM_OUT 0x00000010
#define PCM_IN 0x00000020